### PR TITLE
fix coroutine in serve docs

### LINKED ISSUE #22220: [Bug] Ray serve example from documentation is not working.
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

Ray Serve

### What happened + What you expected to happen

I ran the serve example mentioned in [Getting started guideline](https://docs.ray.io/en/master/ray-overview/index.html).

Error stack 

```
2022-02-08 09:55:20,728 INFO services.py:1374 -- View the Ray dashboard at http://127.0.0.1:8266
(ServeController pid=20004) 2022-02-08 09:55:25,344     INFO checkpoint_path.py:16 -- Using RayInternalKVStore for controller checkpoint and recovery.
(ServeController pid=20004) 2022-02-08 09:55:25,451     INFO http_state.py:98 -- Starting HTTP proxy with name 'SERVE_CONTROLLER_ACTOR:rzEaxa:SERVE_PROXY_ACTOR-node:127.0.0.1-0' on node 'node:127.0.0.1-0' listening on '127.0.0.1:8000'
2022-02-08 09:55:25,783 INFO api.py:475 -- Started Serve instance in namespace 'serve'.
(HTTPProxyActor pid=20011) INFO:     Started server process [20011]
2022-02-08 09:55:26,993 INFO api.py:249 -- Updating deployment 'BoostingModel'. component=serve deployment=BoostingModel
(ServeController pid=20004) 2022-02-08 09:55:27,084     INFO deployment_state.py:920 -- Adding 1 replicas to deployment 'BoostingModel'. component=serve deployment=BoostingModel
2022-02-08 09:55:28,207 INFO api.py:261 -- Deployment 'BoostingModel' is ready at `http://127.0.0.1:8000/iris`. component=serve deployment=BoostingModel
Task Error. Traceback: ray::RayServeWrappedReplica.handle_request() (pid=20009, ip=127.0.0.1)
  File "/Users/gauta/PycharmProjects/RayPy3.9/venv/lib/python3.9/site-packages/ray/serve/utils.py", line 241, in wrap_to_ray_error
    raise exception
  File "/Users/gauta/PycharmProjects/RayPy3.9/venv/lib/python3.9/site-packages/ray/serve/replica.py", line 314, in invoke_single
    result = await method_to_call(*args, **kwargs)
  File "/Users/gauta/PycharmProjects/RayPy3.9/ray_serve.py", line 23, in __call__
    payload = await request.json()["vector"]
TypeError: 'coroutine' object is not subscriptable.
(ServeController pid=20004) 2022-02-08 09:55:28,315     INFO deployment_state.py:940 -- Removing 1 replicas from deployment 'BoostingModel'. component=serve deployment=BoostingModel
(BoostingModel pid=20009) /Users/gauta/PycharmProjects/RayPy3.9/ray_serve.py:23: RuntimeWarning: coroutine 'Request.json' was never awaited
(BoostingModel pid=20009)   payload = await request.json()["vector"]
(BoostingModel pid=20009) RuntimeWarning: Enable tracemalloc to get the object allocation traceback
(venv) 147dda76cba4:RayPy3.9 gauta$ 
```

### Versions / Dependencies

Ray: 1.10.0
Python: 3.9.7
OS: MacOS

### Reproduction script

```
import requests

from sklearn.datasets import load_iris
from sklearn.ensemble import GradientBoostingClassifier

from ray import serve

serve.start()

# Train model.
iris_dataset = load_iris()
model = GradientBoostingClassifier()
model.fit(iris_dataset["data"], iris_dataset["target"])


@serve.deployment(route_prefix="/iris")
class BoostingModel:
    def __init__(self, model):
        self.model = model
        self.label_list = iris_dataset["target_names"].tolist()

    async def __call__(self, request):
        payload = await request.json()["vector"]
        print(f"Received flask request with data {payload}")

        prediction = self.model.predict([payload])[0]
        human_name = self.label_list[prediction]
        return {"result": human_name}


# Deploy model.
BoostingModel.deploy(model)

# Query it!
sample_request_input = {"vector": [1.2, 1.0, 1.1, 0.9]}
response = requests.get(
    "http://localhost:8000/iris", json=sample_request_input)
print(response.text)
```

### Anything else

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by jiaodong**
## Why are these changes needed?

Our homepage docs https://docs.ray.io/en/master/ray-overview/index.html for ray serve is not executable. Issue is same as https://stackoverflow.com/questions/51459793/python-typeerror-coroutine-object-is-not-subscriptable, we need to wrap awaitable first before accessing a key.

## Related issue number

Closes #22220

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/serve/_examples/doc_code/quick_start.py [hunk 1]
```diff
@@ -24,7 +24,7 @@ def __init__(self, model):
         self.label_list = iris_dataset["target_names"].tolist()
 
     async def __call__(self, request):
-        payload = await request.json()["vector"]
+        payload = (await request.json())["vector"]
         print(f"Received flask request with data {payload}")
```
Comment (simon-mo) [inline doc/source/serve/_examples/doc_code/quick_start.py:28]:
```suggestion
        print(f"Received Starlette request with data {payload}")
```

**Review by simon-mo — APPROVED** [review]
We should really test this in python/ray/serve/BUILD. @jiaodong if you don't mind, can you add this to Serve's test suite. Here is how to reference a file in the doc directory

https://github.com/ray-project/ray/blob/master/python/ray/serve/BUILD#L344-L352

**Comment by jiaodong** [comment]
sure, let me give it a try. The doc is `doc/source/ray-overview/index.md` that covers more than just ray serve though

**Comment by jiaodong** [comment]
Hi @maxpumperla just for FYI i think serve's sklearn docs is somewhat duplicated with https://sourcegraph.com/github.com/ray-project/ray/-/blob/python/ray/serve/examples/doc/tutorial_sklearn.py 

Is it possible that we structure it in this way:
1) Each ray library maintains a dir of examples scripts and add to BUILD file on CI to ensure they're runnable on each PR
2) Expose them via py_library as visible bazel target if needed 

Then on the `docs/source` we simply need to reference those files without the need to add new py scripts or test them separately. What do you think ?

**Comment by maxpumperla** [comment]
@jiaodong that's a fantastic idea, thanks for bringing that up. I've already set up `example` folders for the Ray libs I've touched so far. Ideally I'd like these `example` folders to sit in `doc/source` directly.

I'll get back to you on how to set up bazel targets correctly. I'd like to see a setup in which, when our users add new examples, they don't have to understand our CI pipe and touch a `BUILD` file. I.e. all `examples` should be tested by default.

**Comment by jiaodong** [comment]
> @jiaodong that's a fantastic idea, thanks for bringing that up. I've already set up `example` folders for the Ray libs I've touched so far. Ideally I'd like these `example` folders to sit in `doc/source` directly.
> 
> I'll get back to you on how to set up bazel targets correctly. I'd like to see a setup in which, when our users add new examples, they don't have to understand our CI pipe and touch a `BUILD` file. I.e. all `examples` should be tested by default.

That's awesome, thanks for all the work you've done to improve our documentation and glad we both agree ensuring all code snippets we put on ray.io are runnable, secured by CI is desirable.

I agree with "when our users add new examples, they don't have to understand our CI pipeline or touch a BUILD file". One example is @dmatrix is helping to improve Ray Serve documentation https://github.com/ray-project/ray/pull/22401 but it's kinda of a burden in our current setup to 1) Ensure snippets are tested by CI 2) Don't introduce significant friction for people to contribute to our docs. Happy to sync more for better structure to make both 1) and 2) easy :)

**Comment by maxpumperla** [comment]
@jiaodong thanks for your help on that, will definitely take you up on that!

yeah, #22401 is a very good example... in my view there are only two ways in which we should allow code snippets in docs going forward:

1.  Using notebooks. This keeps text and source code naturally together and can be tested as well. I think that's conceptually the easiest way to manage content and should vibe well with our users. For instance, the whole page touched in #22401 (currently `.rst`) could actually just be a big notebook. And to be clear, by notebook I mean any `jupyter` compatible format, but it might really just be an `.ipynb` in the end. This is a primary example for how to do it: https://github.com/ray-project/ray/pull/22315
2. Include snippets. The only other option we have is to use `literalinclude` from a _tested_ script. I think this is mostly attractive for pages in which we have a ton of code `tab`s, but it's more complicated to setup / harder to attract OSS folks who can contribute to it.

In any case, directly including code in a document, unless it's obvious pseudo-code not intended to be run directly, should be disallowed. Those are precisely the examples that don't work on our docs. Luckily, I don't think overall we have too many of them.

**Comment by jiaodong** [comment]
> @jiaodong thanks for your help on that, will definitely take you up on that!
> 
> yeah, #22401 is a very good example... in my view there are only two ways in which we should allow code snippets in docs going forward:
> 
> 1. Using notebooks. This keeps text and source code naturally together and can be tested as well. I think that's conceptually the easiest way to manage content and should vibe well with our users. For instance, the whole page touched in [[DOC] Added a full example how to access deployments  #22401](https://github.com/ray-project/ray/pull/22401) (currently `.rst`) could actually just be a big notebook. And to be clear, by notebook I mean any `jupyter` compatible format, but it might really just be an `.ipynb` in the end. This is a primary example for how to do it: [[docs] [tune] hyperopt notebook #22315](https://github.com/ray-project/ray/pull/22315)
> 2. Include snippets. The only other option we have is to use `literalinclude` from a _tested_ script. I think this is mostly attractive for pages in which we have a ton of code `tab`s, but it's more complicated to setup / harder to attract OSS folks who can contribute to it.
> 
> In any case, directly including code in a document, unless it's obvious pseudo-code not intended to be run directly, should be disallowed. Those are precisely the examples that don't work on our docs. Luckily, I don't think overall we have too many of them.

Using notebook looks much more developer friendly to me :) That was literally what I did when I write my first jobs submission documentation and the original notebook makes it very easy to communicate with our external users. Feel free to grab me if you need anything while refactoring our docs and assign issue/PR to me if they're directly serve docs related.

**Pull request merged**